hstream/src/HStream/Server/Handler.hs

```diff
@@ -0,0 +1,260 @@
+import           HStream.Store
+import qualified HStream.Store.Stream         as S
+import           RIO                          hiding (Handler)
+import           Servant
+import           Servant.Types.SourceT
+import           Z.Data.CBytes                (pack)
+import           Z.Foreign
+
+-------------------------------------------------------------------------
+
+app :: ServerConfig -> IO Application
+app ServerConfig {..} = do
+  s <-
+    State
+      <$> newIORef M.empty
+      <*> newIORef M.empty
+      <*> newIORef []
+      <*> newIORef 0
+      <*> return (pack sLogDeviceConfigPath)
+      <*> mkAdminClient (AdminClientConfig $ pack sLogDeviceConfigPath)
+      <*> mkProducer (ProducerConfig $ pack sLogDeviceConfigPath)
+      <*> return sTopicRepFactor
+  _ <- async $ waitThread s
+  return $ app' s
+
+app' :: State -> Application
+app' s = serve server1API $ hoistServer server1API (liftH s) server1
+
+server1API :: Proxy ServerApi
+server1API = Proxy
+
+server1 :: ServerT ServerApi HandlerM
+server1 = handleTask
+
+handleTask ::
+  HandlerM [TaskInfo]
+    :<|> (ReqSQL -> HandlerM (Either String TaskInfo))
+    :<|> (TaskID -> HandlerM Resp)
+    :<|> (ReqSQL -> HandlerM (SourceIO RecordStream))
+    :<|> (HandlerM Resp)
+handleTask =
+  handleShowTasks
+    :<|> handleCreateTask
+    :<|> handleDeleteTask
+    :<|> handleCreateStreamTask
+    :<|> handleDeleteTaskAll
+
+handleShowTasks :: HandlerM [TaskInfo]
+handleShowTasks = do
+  State {..} <- ask
+  v <- liftIO $ readIORef taskMap
+  return $ fmap snd $ M.elems v
+
+handleQueryTask :: TaskID -> HandlerM (Maybe TaskInfo)
+handleQueryTask t = do
+  State {..} <- ask
+  v <- liftIO $ readIORef taskMap
+  return $ fmap snd $ M.lookup t v
+
+handleDeleteTaskAll :: HandlerM Resp
+handleDeleteTaskAll = do
+  State {..} <- ask
+  ls <- (liftIO $ readIORef waitList)
+  forM_ ls $ \w -> liftIO (cancel w)
+  return $ OK "delete all querys"
+
+handleDeleteTask :: TaskID -> HandlerM Resp
+handleDeleteTask tid = do
+  State {..} <- ask
+  tm <- readIORef taskMap
+  case M.lookup tid tm of
+    Nothing           -> return $ OK "not found the query"
+    Just (Nothing, _) -> return $ OK "query deleted"
+    Just (Just w, _)  -> liftIO (cancel w) >> (return $ OK "delete query")
+
+handleCreateStreamTask :: ReqSQL -> HandlerM (SourceIO RecordStream)
+handleCreateStreamTask (ReqSQL seqValue) = do
+  State {..} <- ask
+  plan' <- liftIO $ try $ streamCodegen seqValue
+  case plan' of
+    Left (e :: SomeException) -> do
+      return $ source [B.pack $ "streamCodegen error: " <> show e]
+    Right plan -> do
+      case plan of
+        SelectPlan sou sink query ->
+          do
+            -----------------------------------
+            tid <- getTaskid
+            time <- liftIO $ getCurrentTime
+
+            let ti = CreateTmpStream tid seqValue sou sink Starting time
+            atomicModifyIORef' taskMap (\t -> (M.insert tid (Nothing, ti) t, ()))
```

Comment:
Extract these operations such as `M.insert` to standalone functions for better readability?